### PR TITLE
New test for installations with existing DDF RAID device

### PR DESCRIPTION
**Pull request by jikortus**
This test is meant to cover https://bugzilla.redhat.com/show_bug.cgi?id=2063791. It only works on RHEL ATM, since the covered bug hasn't been fixed in Rawhide yet.

### CODE REVIEW

**Comment by jstodola** [comment]
Should the test be also excluded from rhel-9 testing due to [bug 2120690](https://bugzilla.redhat.com/show_bug.cgi?id=2120690)?

File: raid-ddf.ks.in [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+cat <<EOF > /usr/libexec/kickstart-test.sh
+# Check that the system booted with the RAID device assembled
+devlink=$(readlink /dev/md/test-raid-ddf)
+if ! lsblk -nl -o MOUNTPOINT /dev/md/${devlink} | grep -q "^/$"; then
+    echo "*** The system hasn't booted with the DDF RAID device /dev/md/test-raid-ddf assembled." > /root/RESULT
+    echo -e "\n\nlsblk output:\n" >> /root/RESULT
+    lsblk -i >> /root/RESULT
+    # Power off the VM, so that the error message is retained in /root/RESULT
+    # (otherwise it would get overwritten after a successful boot)
+    poweroff
```
Comment (rvykydal) [inline raid-ddf.ks.in:43]: I think the poweroff is not needed here as it is actually the successful boot now.

File: raid-ddf.sh [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    echo ${tmpdir}/disk-a.img ${tmpdir}/disk-b.img
+}
+
+additional_runner_args() {
+    # Wait for reboot and shutdown of the VM,
+    # but exit after the specified timeout.
+    echo "--wait $(get_timeout)"
+}
+
+validate() {
```
Comment (rvykydal) [inline raid-ddf.sh:40]: Is this actually needed ? It seems to me that validate() from functions.sh should do the thing.
At least https://github.com/rhinstaller/kickstart-tests/blob/master/rootpw-allow-ssh.sh or https://github.com/rhinstaller/kickstart-tests/blob/master/reboot-initial-setup-gui.sh which check the system after reboot don't need it.

**Comment by rvykydal** [comment]
> Should the test be also excluded from rhel-9 testing due to [bug 2120690](https://bugzilla.redhat.com/show_bug.cgi?id=2120690)?

I think so, it can be done in a way similar to https://github.com/rhinstaller/kickstart-tests/commit/9f9a0df06aefed4a8c6c660c8e0045ff9f5294f2
Also I'd prefer to use the approach with bug reference (https://bugzilla.redhat.com/show_bug.cgi?id=2122327) for disabling on Fedora (instead of skip-on-fedora)

**Comment by rvykydal** [comment]
Thank you for the test!

**Comment by VladimirSlavik** [comment]
Thank you! +1 to everything Radek wrote.

File: raid-ddf.ks.in [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+fi
+
+cat <<EOF > /usr/libexec/kickstart-test.sh
+# Check that the system booted with the RAID device assembled
+devlink=$(readlink /dev/md/test-raid-ddf)
+if ! lsblk -nl -o MOUNTPOINT /dev/md/${devlink} | grep -q "^/$"; then
+    echo "*** The system hasn't booted with the DDF RAID device /dev/md/test-raid-ddf assembled." > /root/RESULT
```
Comment (poncovka) [inline raid-ddf.ks.in:38]: The `kickstart-test.sh` script should only print error messages to stdout if there are any. The `success_on_first_boot.ks` fragment will handle the `/root/RESULT` file and the shutdown of the VM.

**Comment by jikortus** [comment]
Thanks a lot for all of your hints, @jstodola , @rvykydal, @poncovka! I implemented the requested changes, hopefully I haven't missed anything; I hope I understood well how the skips for Rawhide and RHEL-9 were intended.

As for RHEL-9 skipping, there's a strange thing - the test passes (with the RAID device used in the system) in spite of the bug mentioned by Honza. I'm not able to get a pass when doing the installation manually though. This also leads me to a question on how well this covers the actual bug (unless it's some mistake on my side).

**Comment by jstodola** [comment]
Right, the test is passing on RHEL-9 because the RAID array is created and activated manually in %pre kickstart section, so the problem with automatic assembly of the array during boot of the installer will not manifest.
So, it would be better to write the RAID metadata in `prepare_disks()` before starting the installation and get rid of the %pre kickstart section. But you mentioned you hit a problem with this approach, right?

**Comment by jikortus** [comment]
@jstodola, I see - initially I was under the impression that the DDF RAID device was available during installation, but couldn't be assembled in the installed system, now it makes sense to me. Yes, I had issues when I began to implement this test, but those were rather related to my level of knowledge of the kickstart tests environment.

Now I had a second look at the mentioned approach, but still it looks quite difficult to achieve - depending on what the requirements for the test (container) run environment are. It would require a privileged container (or at least have some tweaks exported to the container, such as first creating the disk images outside of the container and 'exporting' them into it as loop devices. The main issue is that mdadm needs a block device and refuses to work with a file.

I also got an idea to create sparse image files, do the RAID setup on them outside of the testing environment and then tar them and include in the test suite as ready-made images for testing. I'll need to see yet if this works or not.

**Comment by jikortus** [comment]
It should be all fine now, I hope :). I ended up using a ready-made images containing the DDF RAID structures and now the test passes (and fails) as expected. The only downside is that the RAID members get synchronized upon installer boot, and therefore the second image (originally sparse) grows to the full size; hopefully this is not a big problem, although this might have happened even with the prior solution, just without me noticing :).

File: raid-ddf.ks.in (hunk 3 above)
Comment (poncovka) [reply raid-ddf.ks.in:31]: Please, remove `> /root/RESULT` and `>> /root/RESULT` from the `kickstart-test.sh` script. It should write error messages to stdout.

Comment (jikortus) [reply raid-ddf.ks.in]: I missed this one, sorry for that. Fixed :).

**Comment by jikortus** [comment]
@rvykydal @poncovka, may I ask you again for a review, please? I hope all of the missing bits have been fixed and the test should be ready to merge...

**Review by poncovka — APPROVED** [review]
I am not sure about the location of the `raid-ddf.tar.xz` file, but that's up to Radek. Otherwise, it looks good to me. Thanks!

File: containers/runner/skip-testtypes [hunk 4]
```diff
@@ -10,10 +10,12 @@ common_skip_array=(
 
 rawhide_skip_array=(
   skip-on-fedora
+  rhbz2122327 # installation with an existing DDF RAID device fails
```
Comment (rvykydal) [inline containers/runner/skip-testtypes:13]: Please move the disabling to the daily_iso_skip_array. We don't actually keep rawhide_skip_array and rawhide_test_skip_array up to date as we are using only the daily_iso_skip_array. Sorry for that it not very obvious.

**Review by rvykydal — CHANGES_REQUESTED** [review]
Looks good to me, just needs disabling of the test on more suitable place.

As for the location of raid-ddf.tar.xz I am not sure where we should store it (perhaps separate images directory.. not sure) so I'll leave moving it for future separate PR.

**Review by rvykydal — APPROVED** [review]
Thank you.

**Comment by jikortus** [comment]
Thank you, @poncovka, @VladimirSlavik, @rvykydal!

**Pull request merged**